### PR TITLE
(newapp) Add `.editorconfig` and `.vscode` extension recomendations

### PR DESCRIPTION
**Pull request by JuanM04**
### What are the changes and their implications?

Since VSCode is the [most used JavaScript editor](https://2020.stateofjs.com/en-US/other-tools/#text_editors), I think we should add [recommended extensions](https://code.visualstudio.com/docs/editor/extension-gallery#_workspace-recommended-extensions) and configuration to new apps.

The extensions are:
- [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)
- [EditorConfig for VSCode](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig)
- [Prettier - Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
- [DotENV](https://marketplace.visualstudio.com/items?itemName=mikestead.dotenv)
- [Babel JavaScript](https://marketplace.visualstudio.com/items?itemName=mgmcdermott.vscode-language-babel)
- [Jest](https://marketplace.visualstudio.com/items?itemName=orta.vscode-jest)
- [Prisma](https://marketplace.visualstudio.com/items?itemName=prisma.prisma)

The default configuration is:
- Format (with Prettier) on Save
- Fix ESLint errors on Save

Also, there is a basic EditorConfig!

### CODE REVIEW

**Comment by flybayer** [comment]
Thanks!

I think we should also hide these files from the list of created files printed during `blitz new`.

**Comment by kevinlangleyjr** [comment]
Just a thought 💡

Maybe we should also add an `.editorconfig` which is usually respected on most IDEs.

**Comment by JuanM04** [comment]
> I think we should also hide these files from the list of created files printed during blitz new.

I didn't find how to do it without harcoding it in the main `Generator` class :confused:

**Comment by flybayer** [comment]
@JuanM04 hardcoding in the generator class is fine I think

**Comment by JuanM04** [comment]
@flybayer I sort of did it, but I found that all the contents from `.vscode` aren't being copied. I've debugging but I don't find the problem. Could you give it a quick look?

**Comment by flybayer** [comment]
@JuanM04 ok sweet. I think dot files aren't copied by default. Because we have the below for gitignore in app-generator.

```
this.fs.move(this.destinationPath("gitignore"), this.destinationPath(".gitignore"))
```

Maybe there's a way to make it copy dot files by default. Otherwise we have to do the above move trick.

**Comment by JuanM04** [comment]
@flybayer what's weird is that `fs.readdir` doesn't detect `.vscode`. In fact, it ignore all the "dot directories". However, when you run a simple node script (or run node from the terminal) that reads the template directory, `.vscode` is listed.

Maybe is something of the local build, but I don't know how to test it

File: packages/generator/src/generator.ts [hunk 1]
```diff
@@ -229,7 +229,7 @@ export abstract class Generator<
 
   async write(): Promise<void> {
     debug("Generator.write...")
-    const paths = readDirRecursive(this.sourcePath(), (name) => {
+    const paths = await readdirRecursive(this.sourcePath(), (name) => {
       const additionalFilesToIgnore = this.filesToIgnore()
       return ![...alwaysIgnoreFiles, ...additionalFilesToIgnore].includes(name)
     })
```
Comment (flybayer) [inline packages/generator/src/generator.ts:235]: Probably the dotfiles are getting filtered by this callback, right?

Comment (JuanM04) [reply packages/generator/src/generator.ts:235]: I've rewritten `readdirRecursive` to check the call to `fs.readdir`, and it doesn't appear. Try adding a `console.log` below `fs.readdir` in `readdir-recursive.ts` and create a new blitz app

**Comment by flybayer** [comment]
@JuanM04 the issue is our [build script](https://github.com/blitz-js/blitz/blob/canary/packages/generator/package.json#L10) doesn't copy the contents of `.vscode`. `packages/generator/dist/templates/app/.vscode/` is empty.

See https://github.com/mysticatea/cpx/issues/34

So we need to workaround this somehow. We could use patch-package to patch cpx.

**Comment by JuanM04** [comment]
> We could use patch-package to patch cpx.

I'll do that. Thank you, I was driving crazy :rofl:

**Comment by JuanM04** [comment]
@flybayer I just [forked `cpx`](https://www.npmjs.com/package/@juanm04/cpx) (which hasn't been updated since 2016 :open_mouth:) and added `--hidden-files`

**Comment by flybayer** [comment]
Sweet, thanks!

**Comment by flybayer** [comment]
Looks like we have a test failure

**Comment by flybayer** [comment]
@JuanM04 the failure introduced by that commit is this:

```
[@blitzjs/cli::jest] FAIL test/commands/new.test.ts (32.455 s)
[@blitzjs/cli::jest]   ● `new` command › when scaffolding new project › pins Blitz to the current version
[@blitzjs/cli::jest] 
[@blitzjs/cli::jest]     TypeError: Cannot read property 'startsWith' of undefined
[@blitzjs/cli::jest] 
[@blitzjs/cli::jest]       1149 | var pluralCamel = /*#__PURE__*/pipe(plural, uncapitalize);
[@blitzjs/cli::jest]       1150 | function camelCaseToKebabCase(transformString) {
[@blitzjs/cli::jest]     > 1151 |   return transformString.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
[@blitzjs/cli::jest]            |                 ^
[@blitzjs/cli::jest]       1152 | }
[@blitzjs/cli::jest]       1153 | 
[@blitzjs/cli::jest]       1154 | var FieldType;
[@blitzjs/cli::jest] 
[@blitzjs/cli::jest]       at AppGenerator.preventFileFromLogging (../generator/dist/blitzjs-generator.cjs.dev.js:1151:17)
[@blitzjs/cli::jest]       at ../generator/dist/blitzjs-generator.cjs.dev.js:763:24
[@blitzjs/cli::jest]           at Array.filter (<anonymous>)
[@blitzjs/cli::jest]       at AppGenerator.run (../generator/dist/blitzjs-generator.cjs.dev.js:759:36)
[@blitzjs/cli::jest]       at New.run (src/commands/new.ts:106:13)
[@blitzjs/cli::jest]       at New._run (../../node_modules/@oclif/command/lib/command.js:43:20)
[@blitzjs/cli::jest]       at whileStayingInCWD (test/commands/new.test.ts:59:13)
[@blitzjs/cli::jest]       at withNewApp (test/commands/new.test.ts:74:13)
[@blitzjs/cli::jest]       at Object.<anonymous> (test/commands/new.test.ts:83:75)
```

And then this commit https://github.com/blitz-js/blitz/pull/2142/commits/105d7200784b2c16652bcdd84508a4860be9685f added another failure which you can solve by mocking out the fs dependencies like we are already doing in `packages/generator/test/generators/app-generator.test.ts`.

**Comment by JuanM04** [comment]
@flybayer I did a small change to `packages/installer/src/executors/new-file-executor.tsx`. It shouldn't break anything, but since I don't know what it does nor how to trigger it, could you check that file (and/or test it manually) for me?

**Comment by flybayer** [comment]
Yay!!

**Pull request merged**